status-handler.js

```diff
@@ -173,6 +182,46 @@ class StatusHandler {
     return true;
   }
 
+  /**
+   * Check if a build is completed for all locales
+   *
+   * @function _isBuildCompleted
+   * @param {Object} data - Message from NSQ
+   * @returns {String} the package name
+   */
+  async _isBuildCompleted(data) {
+    const { version, env, locale } = data;
+    const pkg = this._getPackageName(data);
+
+    const [previousEvents, locales] = await Promise.all([
+      this.models.StatusEvent.findAll({
+        pkg, version, env
+      }),
+      new Promise((resolve, reject) => {
+        this.wrhs.packages.get({ pkg, version, env }, (err, result) => {
+          if (err) return reject(err);
+          resolve(result.extended.locales || []);
+        });
+      })
+    ]);
+
+    const completedLocales = previousEvents
+      .filter(event => event.message === 'carpenterd-worker build completed')
+      .map(event => event.locale);
+
+    const numOfCompletedLocales = (Array.from(new Set(completedLocales))).length;
+
+    // All locales completed and no duplicates
+    if (
+      numOfCompletedLocales === locales.length &&
+      completedLocales.reduce((acc, lc) => acc + (lc === locale), 0) === 1
```

Comment:
Can we not do this type coercion? It makes it really hard to see what you're trying to do here. And even so what you're doing doesn't match what the comment says. if you want to do what the comment says you could just do `completedLocales.length === numOfCompletedLocales`

And if you want to check how many locales match the current locale that came in you should do something like `completedLocales.filter(lc => lc === locale).length === 1` as its infinitely more readable and the extra array will be trivial perf-wise.

Reply:
`numOfCompletedLocales === locales.length` is actually correct. `completedLocales.length === numOfCompletedLocales` won't work. You need to compare with the actual number of locales in the package json.

At the same time, I am not sure what's the problem to use the reduce function. This is a good use case of it. It does the same exact thing you suggest without creating an extra array. Not that perfs are ever going to be an issue here, but still. Also saying "infinitely more readable" seems a bit extreme: it's just a `reduce` function and it should be pretty straight forward for anyone.

Now, since you seems having a strong opinion on this, while I don't, I am going to change this :)